plutus-core/plutus-core/src/Universe/Core.hs

```diff
@@ -0,0 +1,756 @@
+why we have all these 'Typeable' constraints lying around.
+
+Type-wise this is enforced via 'bring' expecting a @uni (T a)@ with @a@ being of kind @Type@.
+I.e. there is no way one could parse a general type and then attempt to bring a constraint for
+that type in scope via 'bring' without first ensuring that the type is of kind star.
+
+Another reason for having the 'Typeable' constraints is that decoding a type application requires
+checking that the expected kind of an argument matches the kind of the provided argument due to us
+having an intrinsically-kinded reprensentation of types.
+-}
+
+{- Note [The TypeApp approach]
+There's an alternative approach to representing universes supporting partial application of
+type constructors. The idea is that we only "shield" those types that are of a non-star kind instead
+of applying @T@ to every single type in the universe. I.e. we specifically embed into @Type@ only
+(possibly partial) type applications and add one more constructor that allows for "running" such
+applications:
+
+    data TypeApp (a :: k)
+
+    data U (a :: Type) where
+        UProtoList  :: U (TypeApp [])
+        UInt        :: U Int
+        UApply      :: U (TypeApp f) -> U a -> U (TypeApp (f a))
+        URunTypeApp :: U (TypeApp a) -> U a
+
+This representation has the advantage of allowing for sharing the universes infrastructure between
+the monomorphic and polymorphic cases. I.e. we could use good old
+
+    data ValueOf uni a = ValueOf (uni a) a
+
+instead of having to use the slightly more awkward
+
+    data ValueOf uni a = ValueOf (uni (T a)) a
+
+One problem that this representation has is redundancy: there are two ways to represent the type
+of lists of integers (for some definition of @SomeTypeIn@):
+
+    SomeTypeIn (URunTypeApp (UProtoList `UApply` UInt))
+    SomeTypeIn (UProtoList `UApply` UInt)
+
+In practice it's not a big deal, we just need not to forget to strip the 'URunTypeApp' constructor
+off whenever that is important (for example, during type normalization).
+
+But a more serious problem with this representation is that we lose all the type safety discussed
+in Note [Decoding universes], which makes it rather easy to shoot oneself in the foot with
+@error "not supposed to happen"@ happening. I.e. we'd need to perform all the same checks
+in client code but without the type checker forcing us to do so like it does now, so that's a huge
+price to pay for some superficial syntactic nicety and hence we choose the safe approach,
+even though that required reworking all the infrastructure in a backwards-incompatible manner.
+-}
+
+-- See Note [Representing polymorphism].
+type T :: forall k. k -> Type
+data T a
+
+-- | Existential quantification as a data type.
+type Some :: forall a. (a -> Type) -> Type
+data Some f = forall x. Some !(f x)
+
+-- | A particular type from a universe.
+type SomeTypeIn :: (Type -> Type) -> Type
+data SomeTypeIn uni = forall k (a :: k). SomeTypeIn !(uni (T a))
```

Comment:
I tried doing that (except with a `newtype` rather than `type`, 'cause I rarely like the idea of exporting a `type`) and I didn't like how it looked or the ergonomics or something. That `Tag` type then has to be of kind `forall k. (Type -> Type) -> k -> Type` and that kind-polymorphism is somewhat unfortunate... Anyway, I don't mind trying again, but I'll use `newtype` then.

Reply:
The idea with `type` is just that it's there for readability rather than anything else, but you could do a newtype also.

Reply:
> you could do a newtype also.

That turned out to be a terrible idea. Reasons:

1. if we adopt using `Tag uni a` in place of `uni (T a)`, we'll end up having nested tagness like here:

```haskell
instance Parsable (SomeTypeIn (Kinded DefaultUni)) where
    parse "bool"       = Just . SomeTypeIn . Tag . Kinded $ Tag DefaultUniBool
```

(it took me several minutes to realize that in a more complex case)

2. we'd have to either duplicate a lot of the API or define type class methods over `Tag uni a` as well, which is somewhat annoying. For example:

```haskell
instance Closed DefaultUni where
    <...>
    bring p (Tag (DefaultUniProtoPair `DefaultUniApply` uniA `DefaultUniApply` uniB)) r =
        bring p (Tag uniA) $ bring p (Tag uniB) r
```

3. the user has to wrap/unwrap `Tag`. For example:

```haskell
instance Closed uni => Flat (SomeTypeIn uni) where
    <...>
    decode = decodeKindedUniFlat <&> \(SomeTypeIn (Tag (Kinded uni))) -> SomeTypeIn uni
```

so that `T` still pops up in user code.

4. but the real problem is that `Tag` is not a sound abstraction: we want to be able to call things like `geq` over it without unpacking it first, but we just can't: `Tag uni a` in its full looks like `Tag @k uni a` where `k` is the kind of `a`, and so if we define `GEq (Tag uni)` that's in fact `GEq (Tag @k uni)`, i.e. it hardcodes a specific kind, while with `GEq uni` we retain the full generality with `T` "escaping" types of arbitrary kinds. Using `Tag` does break our code and if it takes me some effort to deal with that broken abstraction, it's going to be much harder for someone who did not write that code.

I wouldn't mind making `Tag` a type synonym if you wish so, but I consider it pointless. Type synonyms are good when they don't leave their module or when they shorten a definition by a lot or when they are needed to simulate subtyping like in `lens`, but in our case it would be just two different equally short ways to write the same thing with the "undesired" one popping up in error messages etc, so that the user will have to learn what it means anyway.

Reply:
> Type synonyms are good when they don't leave their module or when they shorten a definition by a lot or when they are needed to simulate subtyping like in lens, but in our case it would be just two different equally short ways to write the same thing with the "undesired" one popping up in error messages etc, so that the user will have to learn what it means anyway.

I'm not sure I totally agree with that. I spend a lot of time reading type signatures, and a synonym can be quite helpful there. But I don't mind that much.